packages/react-form-state/README.md

```diff
@@ -0,0 +1,947 @@
+
+#### Example
+
+```typescript
+import {TextField} from '@shopify/polaris';
+import FormState from '@shopify/react-form-state';
+
+function MyComponent() {
+  return (
+    <FormState
+      initialValues={{
+        title: 'Cool title',
+        description: 'Cool product',
+      }}
+    >
+      {formDetails => {
+        const {fields} = formDetails;
+
+        return (
+          <form>
+            <TextField label="Title" {...fields.title} />
+            <TextField label="Description" {...fields.description} />
+          </form>
+        );
+      }}
+    </FormState>
+  );
+}
+```
+
+There is also nothing to stop you from using your own custom inputs.
+
+### Submission
+
+Forms usually need some mechanism of submission. `<FormState />` is not opinionated about how this is done, but does provide a built-in mechanism for managing loading state and updating based on errors from the submission process. Most of this is done through the `onSubmit` property.
+
+```typescript
+  onSubmit={async (fields) => {
+    const result = await updateProduct(fields);
+
+    if (isErrorResult(result)) {
+      return result.errors;
+    }
+  }}
+```
+
+When given an [async](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function) function (or one that returns a [promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function), the `children` function will be rerendered with `submitting: true` until the promise resolves or rejects. Once it has returned the function will be rerendered with `submitting: false` and `errors` either `[]` (if the promise resolved with no value), or with an array of `{field?: string, error: string}` if the promise resolved with an error list.
+
+#### Example
+
+```typescript
+// api.ts
+
+const fakeErrorResult = {error: [{message: 'api is not real'} };
+
+// simple async function that fakes a failing api call
+export function fakeApiCall(data: any) {
+  return new Promise(resolve => {
+    setTimeout(() => {
+      resolve(]);
+    }, 2000);
+  });
+}
+```
+
+```typescript
+// MyPage.ts
+import {fakeApiCall} from './api';
+import {TextField, Button, Banner, FormLayout, Spinner} from '@shopify/polaris';
+import FormState from '@shopify/react-form-state';
+
+export function MyPage() {
+  return (
+    <FormState
+      initialValues={{email: '', password: ''}}
+      onSubmit={async () => {
+        const results = await fakeApiCall();
+        if (results.error) {
+          return results.error;
+        }
+      }}
+    >
+      {({fields, submitting, error, submit}) => {
+        return (
+          <form onSubmit={submit}>
+            {submitting && <Spinner size="large" color="teal" />}
+
+            {errors.length > 0 && (
```

Comment:
should we provide a `hasErrors` prop here? I don't really like the boolean check inside of the jsx like this.

Reply:
oh lol there totally is one `valid`